### PR TITLE
Add option to take screenshot

### PR DESCRIPTION
**Pull request by bryevdv**
@syamajala had asked about a screenshot capability in https://github.com/StanfordLegion/legion/issues/1551#issuecomment-1731977849

This PR adds a screenshot feature to the profiler frontend. Currently it is activated by `<ctrl>+Space` but open to any alternative suggestions (personally prefer keyboard option to a button, but it would be simple to add one). This approach borrows liberally from the [egui plot screenshot example](https://github.com/emilk/egui/pull/2769/files).

Currently just saves the entire window (I think the added context is nice) but image region could be clipped as desired.

foo

cc @manopapad  @elliottslaughter

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -18,8 +18,11 @@ eframe = { version = "0.22.0", default-features = false, features = [
     "default_fonts", # Embed the default egui fonts.
     "glow",          # Use the glow rendering backend. Alternative: "wgpu".
     "persistence",   # Enable restoring app state when restarting the app.
+    "__screenshot", # __screenshot is so we can dump a screenshot using EFRAME_SCREENSHOT_TO
```
Comment (bryevdv) [inline Cargo.toml:21]: This is what all the official `egui` screenshot examples require 🤷

File: src/app.rs [hunk 2]
```diff
@@ -2250,8 +2261,15 @@ impl eframe::App for ProfApp {
         eframe::set_value(storage, eframe::APP_KEY, self);
     }
 
+    fn post_rendering(&mut self, _screen_size_px: [u32; 2], frame: &eframe::Frame) {
+        // this is inspired by the Egui screenshot example
+        if let Some(screenshot) = frame.screenshot() {
+            self.screenshot = Some(screenshot);
+        }
+    }
```
Comment (bryevdv) [inline src/app.rs:2269]: [ref](https://github.com/hacknus/egui/blob/4e7f5870a6d77afb19cbd0ebab1c2c8a712bb41a/crates/eframe/src/epi/mod.rs#L782C1-L783C40):

    /// Request the current frame's pixel data. Needs to be retrieved by calling [`Frame::screenshot`]
    /// during [`App::post_rendering`].

File: src/app.rs [hunk 3]
```diff
@@ -2261,6 +2279,25 @@ impl eframe::App for ProfApp {
             ..
         } = self;
 
+        if cx.take_screenshot {
+            frame.request_screenshot();
+            cx.take_screenshot = false;
+        }
+
+        if let Some(screenshot) = self.screenshot.take() {
+            if let Some(mut path) = rfd::FileDialog::new().save_file() {
```
Comment (bryevdv) [inline src/app.rs:2288]: Currently always defaults to "Untitled" (.png) but we could set a a different default filename if desired

**Comment by lightsighter** [comment]
> Currently it is activated by <ctrl>+Space but open to any alternative suggestions (personally prefer keyboard option to a button, but it would be simple to add one).

Is there a reason not to use the print screen key?  😉

File: src/app.rs (hunk 3 above)
Comment (lightsighter) [reply src/app.rs:2288]: Maybe default to the name-of-the-profile.png?

**Comment by lightsighter** [comment]
> There is no print screen key on Mac keyboards,

On MAC: print-screen == Ctrl+Cmd+3

> also no egui::Key code for print screen that I could see

That's unfortunate.

**Comment by bryevdv** [comment]
> On MAC: print-screen == Ctrl+Cmd+3

Also Shift Command 4, Shift Command 4 + Space, Shift Command 5... We could make any of them work, but I am not sure there is one obvious choice

**Comment by elliottslaughter** [comment]
Definitely *do not* overload the Mac print screen keys. They will work fine as long as we don't mess with them.

I assume Seshu is not on a Mac, so what he needs is for this to be available on other platforms.

File: src/app.rs (hunk 3 above)
Comment (elliottslaughter) [reply src/app.rs:2288]: Or `Screenshot_YYYY-MM-DD-HH-MM-SS.png`, or something like that.

I agree a more informative default name is helpful.

Comment (bryevdv) [reply src/app.rs]: Giver there can be an arbitrary number of source locators I opted for something like @elliottslaughter's datetime suggestion https://github.com/StanfordLegion/prof-viewer/pull/43/commits/3be2ff1a1a31ca761cf39a02079716d8e3265f3b

Screenshot 2024-01-11 at 12 40 52

**Comment by bryevdv** [comment]
@elliottslaughter CI issues seem to be a version incompat
```
error: package `cfg-expr v0.15.6` cannot be built because it requires rustc 1.70.0 or newer, while the currently active rustc version is 1.67.1
```

**Comment by bryevdv** [comment]
@elliottslaughter any thoughts on the CI issue?

**Comment by elliottslaughter** [comment]
I bumped the Rust version, can you merge and try again? (Note: please throw away your `Cargo.lock` changes to avoid downgrading anything by accident.)

**Comment by bryevdv** [comment]
Unfortunately after merge `master` (pushed) the branch no longer builds for me for any obvious reason at a glance. But I'm about to be OOTO so it will be late next week or the week after before I can get back to look more closely.

**Comment by elliottslaughter** [comment]
Something you're doing seems to depend on GTK:

```
The system library `gdk-3.0` required by crate `gdk-sys` was not found.
```

I'm not sure we want to add additional (non-Rust) dependencies? Is there a way to avoid pulling this in?

**Comment by bryevdv** [comment]
no idea offhand, this PR just follows the official `egui` screenshot example code [here](https://github.com/emilk/egui/pull/2769/files) GDK is an image library, though, so perhaps the documented screenshot capability relies on it it somehow.

**Pull request closed without merging**